### PR TITLE
Add a --preserve-it flag

### LINKED ISSUE #170: Significant regression in --fast performance in doctest-0.12 with GHC 8.2.1
(Originally discussed in https://github.com/ekmett/lens/issues/756.)

The `lens` `doctest` test suite began timing out after upgrading to `doctest-0.12`. We discovered that it was due to substantially worse `doctest` performance with GHC 8.2.1 (but not with other GHC versions) when the `--fast` flag is enabled. To reproduce this issue, use this script:

```haskell
-- GenExample.hs
module Main where

import Control.Monad

import System.Environment
import System.Exit
import System.IO

main :: IO ()
main = do
  args <- getArgs
  case args of
    n:_ -> genExamples (read n)
    _ -> do hPutStrLn stderr "usage: runghc GenExamples.hs <num-examples>"
            exitWith $ ExitFailure 1

genExamples :: Int -> IO ()
genExamples nExamples = do
  putStrLn "module Example where"
  ireplicateA_ nExamples genExample

genExample :: Int -> IO ()
genExample i = putStrLn $ unlines
  [ "-- | Simple " ++ exampleName
  , "-- >>> " ++ exampleName
  , "-- 'a'"
  , exampleName ++ " :: Char"
  , exampleName ++ " = 'a'"
  ]
  where
    exampleName :: String
    exampleName = "example" ++ show i

ireplicateA_ :: Applicative m => Int -> (Int -> m a) -> m ()
ireplicateA_ cnt0 f =
    loop cnt0 0
  where
    loop cnt n
        | cnt <= 0  = pure ()
        | otherwise = f n *> (loop (cnt - 1) $! (n + 1))
```

to generate and run an Haskell file with 865 `doctest` examples by doing the following:

```
$ cabal install doctest-0.12 -- Make sure one uses GHC 8.2.1 to build doctest
$ runghc GenExample.hs 865 > Example.hs
$ doctest Example.hs --fast
```

It took me a full three hours to complete all 865 examples. In contrast, without the `--fast` flag:

```
$ time doctest Example.hs
Examples: 865  Tried: 865  Errors: 0  Failures: 0

real    0m8.098s
user    0m8.360s
sys     0m0.672s
```

This slowdown does not occur on `doctest-0.11.4`, regardless of whether the `--fast` flag is used.

### PR DESCRIPTION
**Pull request by RyanGlScott**
Resolves #170 (or at least works around the underlying GHC bug).

### CODE REVIEW

**Comment by sol** [comment]
@RyanGlScott Thanks a lot for looking into this.  I would like this to be `--preserve-it` instead. That is by default `it` should not be preserved.  I think that may cause less headache down the road.

I'm going to deprecate `0.12.0` for now and we will have a new release later on.

**Comment by RyanGlScott** [comment]
@sol, done.

File: src/Options.hs [hunk 1]
```diff
@@ -67,22 +71,34 @@ data Run = Run {
 , runOptions :: [String]
 , runMagicMode :: Bool
 , runFastMode :: Bool
+, runPreserveIt :: Bool
 } deriving (Eq, Show)
 
+defaultMagic, defaultFastMode, defaultPreserveIt :: Bool
```
Comment (sol) [inline src/Options.hs]: To clarify, defining the types of multiple bindings in one line might be convenient for e.g. writing papers, but I personally prefer not to use this particular syntactic feature in real-world code.

**Comment by RyanGlScott** [comment]
I've updated the PR based on your feedback.

File: test/RunSpec.hs [hunk 2]
```diff
@@ -94,9 +97,9 @@ spec = do
         rmDir "test/integration/custom-package-conf/packages/"
         rmDir "test/integration/custom-package-conf/foo/dist/"
 
-  describe "doctestWithFastMode" $ do
+  describe "doctestWithDefaultOptions" $ do
```
Comment (sol) [inline test/RunSpec.hs]: Minor bike shedding, this should be `describe "doctestWithOptions"` now.

Comment (RyanGlScott) [inline test/RunSpec.hs]: Done.

**Pull request merged**